### PR TITLE
MySQL prepared statement support

### LINKED ISSUE #9695: MySQL Workbench queries not logged in Teleport when using Apply
### Description

**What happened**:
"when using MySQL workbench you can edit inline the rows, click apply, it will show you the query that will run, and click to run it. Using that, for some reason is not logging the query in teleport.
if instead of that I just do a regular query without that apply, then it appears as it should."
**What you expected to happen**:
 All MySQL queries should be logged in Teleport. 

### Reproduction Steps
As minimally and precisely as possible, describe step-by-step how to reproduce the problem.
1. Get mysql up and running
2. Set up mysql workbench per https://goteleport.com/docs/database-access/guides/gui-clients/#mysql-workbench
3.  Insert some data manually in the table, you only need to do 1 or 2 rows
4. Open workbench and run a SELECT to populate the result grid
5. Manually edit a cell, and click ‘Apply’. A dialog box will show the UPDATE DML being run

### Server Details
- Teleport version (run `teleport version`):
- Server OS (e.g. from `/etc/os-release`):
- Where are you running Teleport? (e.g. AWS, GCP, Dedicated Hardware):
- Additional details:

### Client Details
- Tsh version (`tsh version`):
- Computer OS (e.g. Linux, macOS, Windows):
- Browser version (for UI-related issues):
- Installed via (e.g. apt, yum, brew, website download):
- Additional details:

### Debug Logs
Please include or attach debug logs, when appropriate. Obfuscate sensitive information!
- Start Teleport with --debug flag (`teleport --debug`)
- Run tsh with --debug flag (`tsh --debug`)

### PR DESCRIPTION
**Pull request by greedy52**
closes #9695
- #9695

## changes

- Adds `COM_STMT_*` parsing in `mysql/protocol`
- Adds `MySQLStatement*` events
- Updates MySQL engine to emit corresponding events
- Updates `webapps` - https://github.com/gravitational/webapps/pull/599

## testing
setup db local proxy and apply a few changes through MySQL workbench as described in the original issue.

Screen Shot 2022-02-10 at 4 23 56 PM

## possible next steps
Parameter values of the prepared statements can be found in the `COM_STMT_EXECUTE` command. Parsing out these values is **NOT** implemented and parameters is always a nil slice in this PR, with the following reasons:

- Not as useful
  - In the reported case of MySQL workbench edit scenario, values are hard coded in the query statement so no parameters in `COM_STMT_EXECUTE`
  - MySQL workbench import or `mysql` client use prepared statement functionalities with `COM_QUERY` (e.g. query = `prepare ...`), instead of dediated `COM_STMT_*` commands
  - Our webapp GUI does not show parameters in the brief description
- Difficulty
  - Parsing parameter values require number of parameters, which is returned from `COM_STMT_PREPARE` response from the server -> we will have to cache parameter info for the statement across commands
  - `COM_STMT_PREPARE` response uses same packet type as a regular `OK` but has different format. `go-mysql/client` is a synchronous client where it waits explicitly for this "special" response after prepare. In contrast, our implementation uses two separate go-routines for each client and server connection, making it difficult to differentiate if an OK response is a normal one or for prepare.

I think it's doable so I can implement this in a separate PR after this basic implementation, or submit an issue to track for later.

### CODE REVIEW

**Review by smallinsky — COMMENTED** [review]
Left some minor comments otherwise LGTM. 
Could you also ensure that MariaDB statements don't diverge from MySQLand this change is also compatible with MariaDB ?

File: lib/srv/db/mysql/protocol/packet.go [hunk 1]
```diff
@@ -144,16 +143,51 @@ func ParsePacket(conn net.Conn) (Packet, error) {
...
 		// https://dev.mysql.com/doc/internals/en/com-change-user.html#packet-COM_CHANGE_USER
-		idx := bytes.IndexByte(packetBytes[5:], 0x00)
+		idx := bytes.IndexByte(packetBytes[packetHeaderAndTypeSize:], 0x00)
 		if idx < 0 {
 			return nil, trace.BadParameter("failed to parse COM_CHANGE_USER packet: %s", packetBytes)
 		}
 		return &ChangeUser{packet: packet, user: string(packetBytes[5 : 5+idx])}, nil
+
+	case mysql.COM_STMT_PREPARE:
+		packet, ok := parseStatementPreparePacket(packet)
```
[the remark this answers is not in the document]
Comment (greedy52) [reply lib/srv/db/mysql/protocol/packet.go:158]: I'd love to refactor existing code to separate functions. Will do that as a separate PR.

**Comment by greedy52** [comment]
@smallinsky All five commands added are compatible with MariaDB. However there are extra command types. Will add [COM_STMT_FETCH](https://mariadb.com/kb/en/com_stmt_fetch/) and [COM_STMT_BULK_FETCH ](https://mariadb.com/kb/en/com_stmt_bulk_execute/)

File: lib/srv/db/mysql/events.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+		UserMetadata:     common.MakeUserMetadata(session),
+		SessionMetadata:  common.MakeSessionMetadata(session),
+		DatabaseMetadata: common.MakeDatabaseMetadata(session),
+		Query:            packet.Query(),
+	}
+}
+
+// makeStatementExecuteEvent creates an audit event for MySQL statement execute
+// command.
+func makeStatementExecuteEvent(session *common.Session, packet *protocol.StatementExecutePacket, parameterDefinitions []mysql.Field) events.AuditEvent {
```
Comment (r0mant) [inline lib/srv/db/mysql/events.go:43]: Looks like `parameterDefinitions` is unused and can be removed.
```suggestion
func makeStatementExecuteEvent(session *common.Session, packet *protocol.StatementExecutePacket) events.AuditEvent {
```

File: lib/srv/db/mysql/events.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	// TODO(greedy52) get parameters from packet and format them for audit.
+	return &events.MySQLStatementExecute{
+		Metadata: common.MakeEventMetadata(session,
+			libevents.DatabaseSessionMySQLStatementExecuteEvent,
+			libevents.MySQLStatementExecuteCode),
+		UserMetadata:     common.MakeUserMetadata(session),
+		SessionMetadata:  common.MakeSessionMetadata(session),
+		DatabaseMetadata: common.MakeDatabaseMetadata(session),
+		StatementID:      packet.StatementID(),
+		Parameters:       nil,
```
Comment (r0mant) [inline lib/srv/db/mysql/events.go:53]:
```suggestion
```

File: lib/srv/db/mysql/events.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		SessionMetadata:  common.MakeSessionMetadata(session),
+		DatabaseMetadata: common.MakeDatabaseMetadata(session),
+		StatementID:      packet.StatementID(),
+		RowsCount:        packet.RowsCount(),
+	}
+}
+
+// makeStatementBulkExecuteEvent creates an audit event for MySQL statement
+// bulk execute command.
+func makeStatementBulkExecuteEvent(session *common.Session, packet *protocol.StatementBulkExecutePacket, parameterDefinitions []mysql.Field) events.AuditEvent {
```
Comment (r0mant) [inline lib/srv/db/mysql/events.go:118]:
```suggestion
func makeStatementBulkExecuteEvent(session *common.Session, packet *protocol.StatementBulkExecutePacket) events.AuditEvent {
```

File: lib/srv/db/mysql/events.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	// TODO(greedy52) get parameters from packet and format them for audit.
+	return &events.MySQLStatementBulkExecute{
+		Metadata: common.MakeEventMetadata(session,
+			libevents.DatabaseSessionMySQLStatementBulkExecuteEvent,
+			libevents.MySQLStatementBulkExecuteCode),
+		UserMetadata:     common.MakeUserMetadata(session),
+		SessionMetadata:  common.MakeSessionMetadata(session),
+		DatabaseMetadata: common.MakeDatabaseMetadata(session),
+		StatementID:      packet.StatementID(),
+		Parameters:       nil,
```
Comment (r0mant) [inline lib/srv/db/mysql/events.go:128]:
```suggestion
```

**Review by r0mant — APPROVED** [review]
lgtm with a few suggestions

File: lib/events/api.go [hunk 6]
```diff
@@ -420,6 +420,31 @@ const (
 	// calls an internal function.
 	DatabaseSessionPostgresFunctionEvent = "db.session.postgres.function"
 
+	// DatabaseSessionMySQLStatementPrepareEvent is emitted when a MySQL client
+	// creates a prepared statement using the parepared statement protocol.
+	DatabaseSessionMySQLStatementPrepareEvent = "db.session.mysql.stmt_prepare"
```
[the remark this answers is not in the document]
Comment (greedy52) [reply lib/events/api.go:425]: sounds good. will update webapp accordingly

**Review by smallinsky — APPROVED** [review]

**Comment by greedy52** [comment]
I found something interesting during testing. Since our server advertises with version `8.0.0-Teleport`, MariaDB clients wouldn't think they are connecting to an "old" version. Even if we do advertise with a higher version, MariaDB server has an "extended" capability flag to specify if it's capable of doing "bulk execute", which we don't advertise either. So with current implementation, we will not receive `COM_STMT_BULK_EXECUTE`. maybe one day

**Pull request merged**